polar-core/src/sugar.rs

```diff
@@ -0,0 +1,850 @@
+    inner: HashMap<Term, Declarations>,
+}
+
+impl Namespaces {
+    pub fn new() -> Self {
+        Self {
+            inner: HashMap::new(),
+        }
+    }
+
+    fn add(&mut self, resource: Term, declarations: Declarations) -> Option<Declarations> {
+        self.inner.insert(resource, declarations)
+    }
+
+    fn remove(&mut self, resource: &Term) -> Option<Declarations> {
+        self.inner.remove(resource)
+    }
+
+    fn exists(&self, resource: &Term) -> bool {
+        self.inner.contains_key(resource)
+    }
+
+    fn get_declaration(&self, resource: &Term, name: &Term) -> PolarResult<&Declaration> {
+        if let Some(declaration) = self.inner[resource].get(name) {
+            Ok(declaration)
+        } else {
+            Err(ParseError::IntegerOverflow {
+                loc: name.offset(),
+                token: format!(
+                    "Undeclared term {} referenced in implication in {} namespace. \
+                        Did you mean to declare it as a role, permission, or relation?",
+                    name.to_polar(),
+                    resource
+                ),
+            }
+            .into())
+        }
+    }
+
+    fn get_related_type(&self, resource: &Term, relation: &Term) -> PolarResult<&Term> {
+        self.get_declaration(resource, relation)?.as_relation()
+    }
+
+    fn local_predicate_name(&self, resource: &Term, name: &Term) -> PolarResult<Symbol> {
+        Ok(self.get_declaration(resource, name)?.as_predicate())
+    }
+
+    fn cross_resource_predicate_name(
+        &self,
+        resource: &Term,
+        relation: &Term,
+        name: &Term,
+    ) -> PolarResult<Symbol> {
+        let related_type = self.get_related_type(resource, relation)?;
+        self.local_predicate_name(related_type, name)
+    }
+}
+
+fn index_declarations(
+    roles: Option<Term>,
+    permissions: Option<Term>,
+    relations: Option<Term>,
+) -> HashMap<Term, Declaration> {
+    // Fold List<role> => HashMap<role, Declaration>
+    let declarations = roles
+        .into_iter()
+        .flat_map(|inner| inner.value().as_list().unwrap().clone())
+        .fold(HashMap::new(), |mut acc, role| {
+            acc.insert(role, Declaration::Role);
+            acc
+        });
+
+    // Fold List<permission> => HashMap<permission_or_role, Declaration>
+    let declarations = permissions
+        .into_iter()
+        .flat_map(|inner| inner.value().as_list().unwrap().clone())
+        .fold(declarations, |mut acc, permission| {
+            acc.insert(permission, Declaration::Permission);
+            acc
+        });
+
+    // Fold Dict<relation, resource> => HashMap<permission_or_role_or_relation, Declaration>
+    relations
+        .into_iter()
+        .flat_map(|inner| inner.value().as_dict().unwrap().fields.clone())
+        .fold(declarations, |mut acc, (relation, resource)| {
+            acc.insert(
+                resource.clone_with_value(value!(relation.0.as_str())),
```

Comment:
The relation comes in as a `Value::Variable` but I need it as a `Value::String` so I can match it against the implication pieces, which are all strings.

Reply:
Why not make it a string in the parser then?

Reply:
Because then it would look like
```rb
relations = { "parent": Org };
```
which doesn't match the syntax for Polar dicts anywhere else in the language. For us, dicts always have `Symbol` keys. If I wanted to parse it as a `Symbol` but then turn it into a `String` unconditionally, I wouldn't be able to use the existing infrastructure we've built for dealing with Polar dictionaries. I'm not sure how onerous that would be, but my first instinct was to reuse the existing stuff.

Reply:
talked through live